google-cloud-spanner/src/main/java/com/google/cloud/spanner/SessionPool.java

```diff
@@ -1696,25 +1716,30 @@ private void handlePrepareSessionFailure(
     synchronized (lock) {
       if (isSessionNotFound(e)) {
         invalidateSession(session);
-      } else if (isDatabaseOrInstanceNotFound(e) || isPermissionDenied(e)) {
-        // Database has been deleted or the user has no permission to write to this database. We
-        // should stop trying to prepare any transactions. Also propagate the error to all waiters,
-        // as any further waiting is pointless.
+      } else if (shouldStopPrepareSessions(e)) {
+        // Database has been deleted or the user has no permission to write to this database, or
+        // there is some other semi-permanent error. We should stop trying to prepare any
+        // transactions. Also propagate the error to all waiters if the database or instance has
+        // been deleted, as any further waiting is pointless.
+        stopAutomaticPrepare = true;
         while (readWriteWaiters.size() > 0) {
           readWriteWaiters.poll().put(e);
         }
         while (readWaiters.size() > 0) {
           readWaiters.poll().put(e);
         }
-        // Remove the session from the pool.
-        allSessions.remove(session);
-        if (isClosed()) {
-          decrementPendingClosures(1);
+        if (isDatabaseOrInstanceNotFound(e)) {
+          // Remove the session from the pool.
+          if (isClosed()) {
+            decrementPendingClosures(1);
+          }
+          allSessions.remove(session);
+          this.resourceNotFoundException =
+              MoreObjects.firstNonNull(
+                  this.resourceNotFoundException, (ResourceNotFoundException) e);
```

Comment:
In the previous logic, we executed this logic also for PERMISSION_DENIED errors. It doesn't anymore. Is this a problem?

Reply:
No, it is still executed for `PERMISSION_DENIED` errors, but that is now covered by the general error handling that also handles other types of errors. The specific errors `DatabaseNotFound` and `InstanceNotFound` are handled specifically, as these should not allow the normal operation of the session pool to be restarted once the error does no longer occur. The reasoning behind that is that a `DatabaseNotFound` exception can only be fixed by creating a new database with the same name as the old one, but that is still a new database.
The other errors, such as `PERMISSION_DENIED` and now also for examle `FAILED_PRECONDITION`, should allow the session pool to restart the automatic preparing of sessions once the error has been fixed. This could be granting the user the necessary permissions or changing the state of database to fulfill all requirements.